ratbag-emu.py

```diff
@@ -26,6 +37,21 @@ def main():
 
 
 if __name__ == '__main__':
+    if sys.version_info < (3, 6):
+        sys.exit('Python 3.6 or later required')
+
+    desc = 'ratbag-emu is a firmware emulator for gaming mice'
+    parser = argparse.ArgumentParser(description=desc)
+    parser.add_argument('-v', '--verbose',
+                        help='Show some debugging informations',
+                        action='store_true',
+                        default=False)
+    ns = parser.parse_args()
+    if ns.verbose:
+        logger.setLevel(logging.DEBUG)
+    else:
+        logger.setLevel(logging.INFO)
```

Comment:
Summary is: you never get enough logs.

The Python logging module makes the debugging much easier by timestamping the logs and adding useful marks. Also it helps a lot having a `--verbose` flag for when you need to see those debug logs.

I think your problem is just that you would rather have the default log for anything not flask related to be debug, and I believe it should be acceptable.

But really, the argument "this will make debugging more frustrating" is not a valid one here because that just means you are not using the correct tools. If you are debugging something on the code, annotate the code where needed or use pdb, but for fetching debug info for other non developer users, we *need* as much debug as possible.

So I am going to rebase @whot's branch and turn his new logger calls into debug so we can move on and nitpick later

Reply:
> But really, the argument "this will make debugging more frustrating" is not a valid one here because that just means you are not using the correct tools. If you are debugging something on the code, annotate the code where needed or use pdb, but for fetching debug info for other non developer users, we need as much debug as possible.

Seeing if a device was added/removed is the exact thing we should use pdb for.

I just want to be able to have an overview of the requests and the raw data sent. Why would pdb make more sense for that? I just want to dump the packets to the console and dissect them later one by one, trying to understand the communication chain. I can't really do that with pdb as we are emulating real hardware, for ex. ratbagd will be expecting replies, I can't just stop the emulator to look at the packet. But I want to be able to do this without all the other logs that are not relevant to me at that time.

Reply:
Maybe we could make the raw communication info and everything else debug? Would that be acceptable?

Reply:
> Would that be acceptable?

just send a PR :)

More seriously, in case you haven't notice we are in the "fixing fires everywhere" state. So it's not very useful to ask for permissions for every changes. For example I just took on me to overwrite @whot's PR and I am sure he won't blame me too much for that. If there is something he doesn't like, well, he can always add more commits later.

So to come back to your question: almost anything is acceptable. However, refusing to include a standard library to handle logs is not. Because I am pretty sure that the library has many more facilities that you know (and that I know), and has been reviewed by a lot of knowledgeable people, and is much better than calling a plain `print()` statement.

Reply:
> For example I just took on me to overwrite @whot's PR and I am sure he won't blame me too much for that.

well, don't take it granted to be allowed overwrite any PR however. This was a minor change and @whot is not in our timezone to fix the changes as they come in. And this PR does a lot of renaming, making it a must to merge it sooner than later.

Reply:
Sure.

Reply:
slightly longer explanation:
- logs are logs and not the same as debugging messages. when you try to narrow down a problem, you add the latter but for logs you just try to get a good subset that helps you narrow down where the issue could be
- connexion may log, but you really want to log at the entry points to your code. doubly so while everything is still changing and you cannot be sure that the next few commits won't accidentally break it. so having something that says "yeah, our code is running now" is always useful 
- for better human-filtering, we could make the log handlers use different colours. we do that for tuhi's interactive shell and it works really well.
- pdb is ... terrible :) well, it works sometimes but a debugger and log files are two orthogonal problems. ideally the log file tells you where you **start** using pdb, as opposed to having to start from `main`
- the generic "filtering is easier than adding what's missing after the fact" is always a good argument :)
- " the other logs that are not relevant to me at that time" is always roughly ~70% of the log, no matter what you log. it's like insurace - you pay a lot but when you need it you get enough back.


might be worth switching default to error and logging everything with info and everything crazy with debug. or figuring out how to add our own levels if we need them, but the brand of the steering wheel doesn't bother me until the car actually drives :)